REGRESSION (252706@main): zillow.com shows a Recaptcha interstitial in private browsing mode

whsieh opened the pull request:
#### 5e4c68005fe4b76b61be2e90bb7064226ce1f69d
<pre>
REGRESSION (252706@main): zillow.com shows a Recaptcha interstitial in private browsing mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=250686">https://bugs.webkit.org/show_bug.cgi?id=250686</a>
rdar://103905006

Reviewed by Youenn Fablet and Ben Nham.

Currently, in private browsing mode (i.e. when using an ephemeral session) `Notification.permission`
unconditionally returns `&quot;denied&quot;` (with `Notification.requestPermission` also unconditionally
denying the request). However, the Permissions API (i.e. `navigator.permissions.query`) implements
inconsistent behavior, and instead returns `&quot;prompt&quot;` in this scenario. On zillow.com, this causes
a bot detection script to erroneously flag the UA as non-human in private browsing, and present
interstitial Recaptcha UI to the user.

Fix this by making the heuristics for &quot;notifications&quot; in `WebPageProxy::queryPermission` consistent
with the Notifications API, such that in an ephemeral session, we return &quot;default&quot; or &quot;prompt&quot;
unless the page has requested access (in which case we&apos;ll deny access, and then subsequently return
&quot;denied&quot;).

Test: http/tests/permissions/permission-state-for-notifications-in-ephemeral-session.html

* LayoutTests/http/tests/permissions/permission-state-for-notifications-in-ephemeral-session-expected.txt:
* LayoutTests/http/tests/permissions/permission-state-for-notifications-in-ephemeral-session.html:
* LayoutTests/http/tests/push-api/permissions-ephemeral-expected.txt:
* LayoutTests/http/tests/push-api/permissions-ephemeral.html:

Rebaseline an existing layout test.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::queryPermission):
* Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
(WebKit::WebNotificationClient::checkPermission):

Canonical link: <a href="https://commits.webkit.org/259040@main">https://commits.webkit.org/259040@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f0122488012be22bca6d89dede9cefa2e099f27f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103565 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12681 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36520 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112801 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173007 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13708 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3580 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95822 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/111973 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109338 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10559 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93625 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38297 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92380 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25222 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79939 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6060 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26624 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6240 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3151 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12222 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46134 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6212 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7993 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -9090,6 +9090,13 @@ void WebPageProxy::queryPermission(const ClientOrigin& clientOrigin, const Permi
+
+        // Always return "denied" in ephemeral browsing sessions, for consistency with
+        // `Notification.permission` and `Notification.requestPermission`.
+        if (sessionID().isEphemeral()) {
```
I think what we want here is something like:
if (sessionID().isEphemeral()) {
     completionHandler(shouldChangeDeniedToPrompt ? PermissionState::Prompt : PermissionState::Denied)
     return;
}
This mimics the fact that the UI delegate would always return denied for ephemeral session.
This would change the above test to expect "prompt".

We would then update the Notification.permission code path to match navigator.permissions.query.
This might mean partially reverting https://github.com/WebKit/WebKit/commit/0570dfb680f59ef121a7faa71d1e8e6297f3594f, in particular the changes in WebNotificationClient::checkPermission.

whsieh replied:
Sounds good — I'll unify the behaviors (but in the opposite direction, so that `Notification.permission` also returns prompt in this case).

whsieh committed:
REGRESSION (252706@main): zillow.com shows a Recaptcha interstitial in private browsing mode
https://bugs.webkit.org/show_bug.cgi?id=250686
rdar://103905006

Reviewed by Youenn Fablet and Ben Nham.

Currently, in private browsing mode (i.e. when using an ephemeral session) `Notification.permission`
unconditionally returns `"denied"` (with `Notification.requestPermission` also unconditionally
denying the request). However, the Permissions API (i.e. `navigator.permissions.query`) implements
inconsistent behavior, and instead returns `"prompt"` in this scenario. On zillow.com, this causes
a bot detection script to erroneously flag the UA as non-human in private browsing, and present
interstitial Recaptcha UI to the user.

Fix this by making the heuristics for "notifications" in `WebPageProxy::queryPermission` consistent
with the Notifications API, such that in an ephemeral session, we return "default" or "prompt"
unless the page has requested access (in which case we'll deny access, and then subsequently return
"denied").

Test: http/tests/permissions/permission-state-for-notifications-in-ephemeral-session.html

* LayoutTests/http/tests/permissions/permission-state-for-notifications-in-ephemeral-session-expected.txt:
* LayoutTests/http/tests/permissions/permission-state-for-notifications-in-ephemeral-session.html:
* LayoutTests/http/tests/push-api/permissions-ephemeral-expected.txt:
* LayoutTests/http/tests/push-api/permissions-ephemeral.html:

Rebaseline an existing layout test.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::queryPermission):
* Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
(WebKit::WebNotificationClient::checkPermission):

Canonical link: https://commits.webkit.org/259040@main
